src/Compilers/CSharp/Test/Semantic/Semantics/StructConstructorTests.cs

```diff
@@ -835,6 +841,772 @@ .maxstack  1
+                // (20,12): error CS0171: Field 'S2.x' must be fully assigned before control is returned to the caller
+                //     public S2() { }
+                Diagnostic(ErrorCode.ERR_UnassignedThis, "S2").WithArguments("S2.x").WithLocation(20, 12),
+                // (29,12): error CS0171: Field 'S3.x' must be fully assigned before control is returned to the caller
+                //     public S3()
+                Diagnostic(ErrorCode.ERR_UnassignedThis, "S3").WithArguments("S3.x").WithLocation(29, 12));
+        }
+
+        [Theory]
+        [InlineData(LanguageVersion.CSharp10)]
+        [InlineData(null)]
+        public void DefaultThisInitializer_07(LanguageVersion? languageVersion)
+        {
+            var source =
+@"#pragma warning disable 169
+#pragma warning disable 414
+struct S1
+{
+    int x = 1;
+    int y;
+    public S1(int y) : this()
+    {
+    }
+    public S1() { }
+}
+struct S2
+{
+    int x = 2;
+    int y;
+    public S2(int y) : this()
+    {
+        this.y = y;
+    }
+    public S2() { }
+}
+struct S3
+{
+    int x = 3;
+    int y;
+    public S3(int y) : this()
+    {
+    }
+    public S3()
+    {
+        this.y = -3;
+    }
+}";
+
+            var comp = CreateCompilation(source, parseOptions: GetParseOptions(languageVersion));
+            comp.VerifyDiagnostics(
+                // (10,12): error CS0171: Field 'S1.y' must be fully assigned before control is returned to the caller
+                //     public S1() { }
+                Diagnostic(ErrorCode.ERR_UnassignedThis, "S1").WithArguments("S1.y").WithLocation(10, 12),
+                // (20,12): error CS0171: Field 'S2.y' must be fully assigned before control is returned to the caller
+                //     public S2() { }
+                Diagnostic(ErrorCode.ERR_UnassignedThis, "S2").WithArguments("S2.y").WithLocation(20, 12));
+        }
+
+        [Theory]
+        [InlineData(LanguageVersion.CSharp10)]
+        [InlineData(null)]
+        public void DefaultThisInitializer_08(LanguageVersion? languageVersion)
+        {
+            var source =
+@"struct S1
+{
+    int x1 = y1 + 1;
+    int y1;
+    public S1(int y)
+    {
+        this.y1 = y;
+    }
+}
+record struct R1
+{
+    int x1 = y1 + 1;
+    int y1;
+    public R1(int y)
+    {
+        this.y1 = y;
+    }
+}";
+
+            var comp = CreateCompilation(source, parseOptions: GetParseOptions(languageVersion));
+            comp.VerifyDiagnostics(
+                // (3,14): error CS0236: A field initializer cannot reference the non-static field, method, or property 'S1.y1'
+                //     int x1 = y1 + 1;
+                Diagnostic(ErrorCode.ERR_FieldInitRefNonstatic, "y1").WithArguments("S1.y1").WithLocation(3, 14),
+                // (3,14): error CS0170: Use of possibly unassigned field 'y1'
+                //     int x1 = y1 + 1;
+                Diagnostic(ErrorCode.ERR_UseDefViolationField, "y1").WithArguments("y1").WithLocation(3, 14),
```

Comment:
It looks like there are 5 or so tests that will be affected. But it might simpler if I leave these combined for now, then split them when merging into main.